[quant][pt2e] Add reference representation for quantized adaptive_avg_pool2d

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105784
* #105783
* __->__ #105709
* #105708

Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_adaptive_avg_pool2d

Although right now it is not really testing things since there is some problem with dynamo export

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] Add reference representation for quantized adaptive_avg_pool2d

Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_adaptive_avg_pool2d

Although right now it is not really testing things since there is some problem with dynamo export

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized adaptive_avg_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_adaptive_avg_pool2d

Although right now it is not really testing things since there is some problem with dynamo export

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -131,6 +131,33 @@ def _reference_quantized_max_pool2d(x_i8, x_scale, x_zero_point, out_scale, out_
+    output_size = (1, 1)
+    x_i32 = x_i8.to(torch.int32)
+    # TODO: use mul.Scalar, need to change how literal args are handled
+    x_i32 = out_dtype(torch.ops.aten.mul.Tensor, torch.int32, x_i32, (x_scale / out_scale))
+    acc_i32 = torch.ops.aten._adaptive_avg_pool2d(x_i32, output_size)
+    acc_i32 = acc_i32 - x_zero_point * x_scale / out_scale + out_zero_point # int32 constant
```
As my comment on maxpool2d diff, this should be rewritten such that
1. zero point subtract
2. adaptive_avg_pool
3. rescale with x_cale/out_scale
4. add zero point.

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized adaptive_avg_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_adaptive_avg_pool2d

Although right now it is not really testing things since there is some problem with dynamo export

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized adaptive_avg_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_adaptive_avg_pool2d

Although right now it is not really testing things since there is some problem with dynamo export

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized adaptive_avg_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_adaptive_avg_pool2d

Although right now it is not really testing things since there is some problem with dynamo export

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1920,6 +1931,22 @@ def test_representation_maxpool2d(self):
+
+        example_inputs = (torch.randn(1, 3, 3, 3),)
+
+        self._test_representation(
```
My concern with all the rewrites is testing. I remember you said dynamo export is timing out. Do you know the issue? Else we are landing stuff that may not be functionally correct

jerryzh168 replied:
I have tested these locally, yeah I plan to have a repro soon

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized adaptive_avg_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_adaptive_avg_pool2d

Although right now it is not really testing things since there is some problem with dynamo export

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized adaptive_avg_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_adaptive_avg_pool2d

Although right now it is not really testing things since there is some problem with dynamo export

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
